### PR TITLE
Fixes to lib/hash and lib/big_int_util.

### PR DESCRIPTION
**Pull request by fruffy**
A couple fixes after #4623 has been merged.

- Move the boost definition of `boost::multiprecision::cpp_int` into its own header file.
- Fix missing of typing for BoolLiterals with SourceInfo.
- Move `big_int` hashing to `big_int_utils`.

### CODE REVIEW

**Comment by fruffy** [comment]
There is a new spurious failure in the Bazel build... it looks like it is missing the `HAVE_LIBBOOST_GRAPH` definition.

File: BUILD.bazel
Comment (vlstill) [inline BUILD.bazel:1]: I have no ability to check the bazel config.

File: lib/alloc_trace.cpp [hunk 1]
```diff
@@ -64,9 +64,6 @@ void AllocTrace::count(void **bt, size_t sz) {
 }
 
 std::ostream &operator<<(std::ostream &out, const AllocTrace &at) {
-#if HAVE_LIBGC
-    PauseTrace temp_pause;
-#endif
```
Comment (vlstill) [inline lib/alloc_trace.cpp:69]: Why is this removed? I don't know what it does, but it does not seem very related.

**Review by vlstill — COMMENTED** [review]
Just one question. The rest look OK to me.

File: BUILD.bazel
Comment (fruffy) [reply BUILD.bazel:1]: I am just guessing here but it does seem to fix the spurious failure.

File: lib/alloc_trace.cpp (hunk 1 above)
Comment (fruffy) [reply lib/alloc_trace.cpp:69]: A warning popped out https://github.com/p4lang/p4c/actions/runs/9003590407/job/24734561146. Not sure what triggered it to appear only now but this piece of code does seem unused.

Comment (vlstill) [reply lib/alloc_trace.cpp:69]: Weird. I would expect it to be a guard that does something in its constructor and destructor (based on the name).

Comment (fruffy) [reply lib/alloc_trace.cpp:69]: I can alternatively just add a NOLINT.

Comment (vlstill) [reply lib/alloc_trace.cpp:69]: Indeed, see https://github.com/p4lang/p4c/blob/main/lib/alloc_trace.h#L67. I don't think this should be removed. Is it possible this build (which produced the warning) is without GC? I think the "right" solution to suppress the warning might be to put and explicit destructor there even for the case where the `HAVE_LIBGC` is false.

Comment (fruffy) [reply lib/alloc_trace.cpp]: `alloc_trace.h` did not include config and removing `config.h` from `big_int_util.h` caused the `#define` to be missing.

**Review by vlstill — APPROVED** [review]
